src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/Datasource.java

```diff
@@ -0,0 +1,487 @@
+                Instant lastFailedAt = args[2] == null ? null : Instant.ofEpochMilli((long) args[2]);
+                Instant lastSkippedAt = args[3] == null ? null : Instant.ofEpochMilli((long) args[3]);
+                return new UpdateStats(lastSucceededAt, lastProcessingTimeInMillis, lastFailedAt, lastSkippedAt);
+            }
+        );
+
+        static {
+            PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_SUCCEEDED_AT_FIELD);
+            PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_PROCESSING_TIME_IN_MILLIS_FIELD);
+            PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_FAILED_AT_FIELD);
+            PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_SKIPPED_AT);
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject();
+            if (lastSucceededAt != null) {
+                builder.timeField(
+                    LAST_SUCCEEDED_AT_FIELD.getPreferredName(),
+                    LAST_SUCCEEDED_AT_FIELD_READABLE.getPreferredName(),
+                    lastSucceededAt.toEpochMilli()
+                );
+            }
+            if (lastProcessingTimeInMillis != null) {
+                builder.field(LAST_PROCESSING_TIME_IN_MILLIS_FIELD.getPreferredName(), lastProcessingTimeInMillis);
+            }
+            if (lastFailedAt != null) {
+                builder.timeField(
+                    LAST_FAILED_AT_FIELD.getPreferredName(),
+                    LAST_FAILED_AT_FIELD_READABLE.getPreferredName(),
+                    lastFailedAt.toEpochMilli()
+                );
+            }
+            if (lastSkippedAt != null) {
+                builder.timeField(
+                    LAST_SKIPPED_AT.getPreferredName(),
+                    LAST_SKIPPED_AT_READABLE.getPreferredName(),
+                    lastSkippedAt.toEpochMilli()
+                );
+            }
+            builder.endObject();
+            return builder;
+        }
+    }
+
+    /**
+     * Builder class for Datasource
+     */
+    public static class Builder {
+        public static Datasource build(final PutDatasourceRequest request) {
+            String id = request.getDatasourceName();
+            Instant lastUpdateTime = Instant.now();
+            Instant enabledTime = null;
+            boolean isEnabled = false;
+            Schedule schedule = new IntervalSchedule(Instant.now(), (int) request.getUpdateIntervalInDays().days(), ChronoUnit.DAYS);
+            String endpoint = request.getEndpoint();
+            DatasourceState state = DatasourceState.PREPARING;
+            List<String> indices = new ArrayList<>();
+            Database database = new Database(null, null, null, null, null);
+            UpdateStats updateStats = new UpdateStats(null, null, null, null);
```

Comment:
if all the parameters to constructor are null then why cannot we create a noArgsConstructor?

Reply:
@heemin32